### PR TITLE
pkp/pkp-lib#6970 Laravel event discovery implementation

### PR DESCRIPTION
**Pull request by Vitaliy-1**

### CODE REVIEW

File: classes/core/PKPEventServiceProvider.inc.php [hunk 1]
```diff
@@ -78,6 +84,82 @@ public function boot()
...
+        {
+            /**
+             * @param SplFileInfo $file
+             * @param string $basePath base path of the application
+             * @return string listener's full class name
+             */
+            protected static function classFromFile(SplFileInfo $file, $basePath): string
+            {
+                $pathFromBase = trim(Str::replaceFirst($basePath, '', $file->getRealPath()), DIRECTORY_SEPARATOR);
+                $libPath = 'lib' . DIRECTORY_SEPARATOR . 'pkp' . DIRECTORY_SEPARATOR;
```
Comment (asmecher) [inline classes/core/PKPEventServiceProvider.inc.php:129]: I think it's safe and way more readable to just use `/` in place of `DIRECTORY_SEPARATOR.

File: classes/core/PKPEventServiceProvider.inc.php [hunk 2]
```diff
@@ -78,6 +84,82 @@ public function boot()
...
+
+    /**
+     * Get the listener directories that should be used to discover events.
+     *
+     * @return array
+     */
+    protected function discoverEventsWithin()
+    {
+        return [
+            $this->app->basePath('lib/pkp/classes/observers/listeners'),
```
Comment (asmecher) [inline classes/core/PKPEventServiceProvider.inc.php:159]: It's arguable whether we should have this inside the `classes` directories or not, but until the autoloader supports that I can see why you've put them here :)

I'll continue to think about that and hope to have a proposal next week.

File: classes/core/PKPEventServiceProvider.inc.php [hunk 3]
```diff
@@ -78,6 +84,82 @@ public function boot()
...
+                );
+            }, []);
+    }
+
+    /**
+     * Get the listener directories that should be used to discover events.
+     *
+     * @return array
+     */
+    protected function discoverEventsWithin()
```
Comment (asmecher) [inline classes/core/PKPEventServiceProvider.inc.php:156]: Does this scan happen on a per-request basis or does this use the Laravel event cache? (https://laravel.com/docs/8.x/events#event-discovery-in-production)

File: classes/core/PKPEventServiceProvider.inc.php (hunk 1 above)
Comment (Vitaliy-1) [reply classes/core/PKPEventServiceProvider.inc.php:129]: I'm not sure here what `SplFileInfo::getRealPath()` returns on the Windows machine. `DIRECTORY_SEPARATOR` here is for correct matching when doing replacements. According to the comment here it returns backslashes: https://www.php.net/manual/en/splfileinfo.getpathname.php

File: classes/core/PKPEventServiceProvider.inc.php (hunk 3 above)
Comment (Vitaliy-1) [reply classes/core/PKPEventServiceProvider.inc.php:156]: No, it requires registration of at least Laravel's `FileServiceProvider`. Haven't dive deep into the code, but maybe `CacheServiceProvider` also. I'm not sure how slow the iteration [through listener's methods](https://github.com/laravel/framework/blob/cada3d21c7f127ba01bb949b2fa045b7faaedbca/src/Illuminate/Foundation/Events/DiscoverEvents.php#L55) with `ReflectionClass` is. Do you think event caching is something that should be implemented right away?

Comment (Vitaliy-1) [reply classes/core/PKPEventServiceProvider.inc.php:156]: On the second look, I can override Laravel's methods to use our file cache. Not necessary to integrate additional Laravel services. @asmecher, do you think it's worth an effort to do it now, before merging this PR?

Comment (asmecher) [reply classes/core/PKPEventServiceProvider.inc.php:156]: I'm content to start as it is, but we'll have to watch as we adopt the event toolset to see whether performance becomes an issue. Thanks!

**Pull request merged**